File: MPP-lab-project/src/Repository/DBRepository/SwimmerRaceDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using log4net;
using MPP_lab_project.Domain;
using MPP_lab_project.Repository.Interfaces;
using MPP_lab_project.Utils;

namespace MPP_lab_project.Repository.DBRepository;

public class SwimmerRaceDBRepository : ISwimmerRaceRepository
{
    private static readonly ILog Logger = LogManager.GetLogger("RaceDbRepository");
    IDictionary<String, String> properties;
    
    public ISwimmerRepository SwimmerRepository { get; set; }
    public IRaceRepository RaceRepository { get; set; }

    public SwimmerRaceDBRepository(ISwimmerRepository swimmerRepository, IRaceRepository raceRepository, IDictionary<string, string> properties)
    {
        Logger.InfoFormat("Initialising RaceDBRepository...");
        this.properties = properties;
        SwimmerRepository = swimmerRepository;
        RaceRepository = raceRepository;
    }

    public int Add(SwimmerRace elem)
    {
        Logger.InfoFormat("Add(swimmer = {0})", elem);
        
        int id = -1;
        
        IDbConnection connection = DbUtils.GetConnection(properties);
        using (IDbCommand comm = connection.CreateCommand())
        {
            comm.CommandText = "insert into SwimmersRaces (id_swimmer, id_race) values (@idSwimmer, @idRace) returning id;";
            
            IDbDataParameter paramIDSwimmer = comm.CreateParameter();
            paramIDSwimmer.ParameterName = "@idSwimmer";
            paramIDSwimmer.Value = elem.Swimmer.ID;
            comm.Parameters.Add(paramIDSwimmer);
            
            IDbDataParameter paramIDRace = comm.CreateParameter();
            paramIDRace.ParameterName = "@idRace";
            paramIDRace.Value = elem.Race.ID;
            comm.Parameters.Add(paramIDRace);

            id = Convert.ToInt32(comm.ExecuteScalar());
        }
        
        Logger.InfoFormat("Result: id = {0}", id);
        return id;
    }

    public void Delete(SwimmerRace elem)
    {
        throw new System.NotImplementedException();
    }

    public void Update(SwimmerRace elem, int id)
    {
        throw new System.NotImplementedException();
    }

    public SwimmerRace FindById(int id)
    {
        throw new System.NotImplementedException();
    }

    public int GetNumberOfSwimmersForRace(Race race)
    {
        Logger.InfoFormat("GetNumberOfSwimmersForRace(race = {0})", race);

        int noSwimmers = 0;
        
        IDbConnection connection = DbUtils.GetConnection(properties);
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText = "select count(*) from SwimmersRaces where id_race = @idRace;";

            IDbDataParameter paramIDRace = command.CreateParameter();
            paramIDRace.ParameterName = "idRace";
            paramIDRace.Value = race.ID;
            command.Parameters.Add(paramIDRace);
            
            using (IDataReader dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    noSwimmers = dataReader.GetInt32(0);
                }
            }
        }
        
        Logger.InfoFormat("Result: numberOfSwimmers = {0}", noSwimmers);
        return noSwimmers;
    }

    public List<Swimmer> FindAllSwimmersForRace(Race race)
    {
        Logger.InfoFormat("FindAllSwimmersForRace(race = {0})", race);

        List<Swimmer> swimmers = new List<Swimmer>();
        
        IDbConnection connection = DbUtils.GetConnection(properties);
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText =
                "select S.id, S.firstName, S.lastName, S.age from Swimmers S inner join SwimmersRaces SR on S.id = SR.id_swimmer where SR.id_race = ?;";

            IDbDataParameter paramIDRace = command.CreateParameter();
            paramIDRace.ParameterName = "idRace";
            paramIDRace.Value = race.ID;
            command.Parameters.Add(paramIDRace);
            
            using (IDataReader dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    int id = dataReader.GetInt32(0);
                    String firstName = dataReader.GetString(1);
                    String lastName = dataReader.GetString(2);
                    int age = dataReader.GetInt32(3);
                    swimmers.Add(new Swimmer(id, firstName, lastName, age));
                }
            }
        }
        
        Logger.InfoFormat("Result: allSwimmersForRace = {0}", swimmers);
        return swimmers;
    }

    public List<Race> FindAllRacesForSwimmer(Swimmer swimmer)
    {
        Logger.InfoFormat("FindAllRacesForSwimmer(swimmer = {0})", swimmer);

        List<Race> allRaces = new List<Race>();
        
        IDbConnection connection = DbUtils.GetConnection(properties);
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText =
                "select R.id, R.distance, R.style, R.swimmersNumber from main.Races R inner join SwimmersRaces SR on R.id = SR.id_race where SR.id_swimmer = @idSwimmer;";

            IDbDataParameter paramIDSwimmer = command.CreateParameter();
            paramIDSwimmer.ParameterName = "idSwimmer";
            paramIDSwimmer.Value = swimmer.ID;
            command.Parameters.Add(paramIDSwimmer);
            
            using (IDataReader dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    int id = dataReader.GetInt32(0);
                    SwimmingDistances swimmingDistances =
                        SwimmingDistancesMethods.DistanceFromInteger(dataReader.GetInt32(1));
                    SwimmingStyles swimmingStyles = SwimmingStylesMethods.StyleFromInteger(dataReader.GetInt32(2));
                    int swimmersNumber = dataReader.GetInt32(3);
                    allRaces.Add(new Race(id, swimmingDistances, swimmingStyles, swimmersNumber));
                }
            }
        }
        
        Logger.InfoFormat("Result: allRacesForSwimmer = {0}", allRaces);
        return allRaces;
    }
}

File: swimming-contest-software/src/main/java/swimmingApp/repository/dbRepository/SwimmerRaceDBRepository.java
package swimmingApp.repository.dbRepository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import swimmingApp.domain.entities.Race;
import swimmingApp.domain.entities.Swimmer;
import swimmingApp.domain.entities.SwimmerRace;
import swimmingApp.domain.enums.SwimmingDistances;
import swimmingApp.domain.enums.SwimmingStyles;
import swimmingApp.repository.interfaces.RaceRepository;
import swimmingApp.repository.interfaces.SwimmerRaceRepository;
import swimmingApp.repository.interfaces.SwimmerRepository;
import swimmingApp.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class SwimmerRaceDBRepository implements SwimmerRaceRepository {

    private final JdbcUtils jdbcUtils;
    private static final Logger logger = LogManager.getLogger();
    private SwimmerRepository swimmerRepository;
    private RaceRepository raceRepository;

    public SwimmerRaceDBRepository(Properties properties, SwimmerRepository swimmerRepository, RaceRepository raceRepository) {
        logger.info("Initialising SwimmerRaceDBRepository with properties {}", properties);
        jdbcUtils = new JdbcUtils(properties);
        this.swimmerRepository = swimmerRepository;
        this.raceRepository = raceRepository;
    }

    public void setSwimmerRepository(SwimmerRepository swimmerRepository) {
        this.swimmerRepository = swimmerRepository;
    }

    public void setRaceRepository(RaceRepository raceRepository) {
        this.raceRepository = raceRepository;
    }

    @Override
    public Integer add(SwimmerRace elem) {
        logger.traceEntry("add(SwimmerRace = {})", elem);
        Integer id = null;

        Connection connection = jdbcUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "insert into main.SwimmersRaces (id_swimmer, id_race) values (?, ?);"
        ))
        {
            preparedStatement.setInt(1, elem.getSwimmer().getID());
            preparedStatement.setInt(2, elem.getRace().getID());
            int result = preparedStatement.executeUpdate();
            ResultSet resultSet = preparedStatement.getGeneratedKeys();
            resultSet.next();
            id = resultSet.getInt(1);

        }
        catch (SQLException e) {
            e.printStackTrace();
        }
        return logger.traceExit("Result: id = {}", id);
    }

    @Override
    public void delete(SwimmerRace elem) {

    }

    @Override
    public void update(SwimmerRace elem, Integer id) {

    }

    @Override
    public SwimmerRace findById(Integer id) {
        return null;
    }

    @Override
    public Integer getNumberOfSwimmersForRace(Race race) {
        logger.traceEntry("getNumberOfSwimmersForRace(race = {})", race);

        int numberOfSwimmers = 0;

        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "select count(*) from main.SwimmersRaces where id_race = ?;"
        )) {
            preparedStatement.setInt(1, race.getID());
            ResultSet resultSet = preparedStatement.executeQuery();
            resultSet.next();
            numberOfSwimmers = resultSet.getInt(1);
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return logger.traceExit("Result: numberOfSwimmers = {}", numberOfSwimmers);
    }

    @Override
    public List<Swimmer> findAllSwimmersForRace(Race race) {
        logger.traceEntry("getNumberOfSwimmersForRace(race = {})", race);
        List<Swimmer> allSwimmersForRace = new ArrayList<>();

        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "select S.id, S.firstName, S.lastName, S.age from main.Swimmers S inner join SwimmersRaces SR on S.id = SR.id_swimmer where SR.id_race = ?;"
        )) {
            preparedStatement.setInt(1, race.getID());
            ResultSet resultSet = preparedStatement.executeQuery();
            while (resultSet.next()) {
                Integer id = resultSet.getInt(1);
                String firstName = resultSet.getString(2);
                String lastName = resultSet.getString(3);
                Integer age = resultSet.getInt(4);
                Swimmer swimmer = new Swimmer(id, firstName, lastName, age);
                allSwimmersForRace.add(swimmer);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return logger.traceExit("Result: allSwimmersForRace = {}", allSwimmersForRace);
    }

    @Override
    public List<Race> findAllRacesForSwimmer(Swimmer swimmer) {
        logger.traceEntry("findAllRacesForSwimmer(swimmer = {})", swimmer);

        List<Race> allRacesForSwimmer = new ArrayList<>();

        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "select R.id, R.distance, R.style, R.swimmersNumber from main.Races R inner join SwimmersRaces SR on R.id = SR.id_race where SR.id_swimmer = ?;"
        )) {
            preparedStatement.setInt(1, swimmer.getID());
            ResultSet resultSet = preparedStatement.executeQuery();
            while (resultSet.next()) {
                Integer id = resultSet.getInt(1);
                Integer distance = resultSet.getInt(2);
                Integer style = resultSet.getInt(3);
                Integer age = resultSet.getInt(4);
                Race race = new Race(id,
                        SwimmingDistances.distanceFromInteger(distance),
                        SwimmingStyles.styleFromInteger(style),
                        age);
                allRacesForSwimmer.add(race);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return logger.traceExit("Result: allRacesForSwimmer = {}", allRacesForSwimmer);
    }
}
